### PR TITLE
BabylonNative/InputManager: Prep BJS Input to connect with Babylon Native

### PR DESCRIPTION
**Pull request by PolygonalSun**
There are several changes to files that will allow BabylonNative to talk with Babylon.js' InputManager.  This change alone will not provide this functionality and requires additional files on the Native side.  It should be noted that this is built against 5.0 alpha-55, as that is the current version used by Babylon Native as of the creation of the PR.

Changes have been made to the InputManager, Cameras, and Scene Helper only.  No external tools were updated yet.

### CODE REVIEW

File: src/Cameras/Inputs/BaseCameraPointersInput.ts [hunk 1]
```diff
@@ -52,6 +53,7 @@ export abstract class BaseCameraPointersInput implements ICameraInput<Camera> {
         const element = engine.getInputElement();
         var previousPinchSquaredDistance = 0;
         var previousMultiTouchPanPosition: Nullable<PointerTouch> = null;
+        if (!engine.getHostDocument()) this._usingNative = true;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/BaseCameraPointersInput.ts]: Removed checks for native

File: src/Cameras/Inputs/flyCameraMouseInput.ts [hunk 2]
```diff
@@ -82,6 +83,7 @@ export class FlyCameraMouseInput implements ICameraInput<FlyCamera> {
     public attachControl(noPreventDefault?: boolean): void {
         noPreventDefault = Tools.BackCompatCameraNoPreventDefault(arguments);
         this.noPreventDefault = noPreventDefault;
+        if (!this.camera.getEngine().getHostDocument()) this._usingNative = true;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/flyCameraMouseInput.ts]: Removed checks for native

File: src/Cameras/Inputs/BaseCameraPointersInput.ts [hunk 3]
```diff
@@ -99,9 +101,9 @@ export abstract class BaseCameraPointersInput implements ICameraInput<Camera> {
                 this.onTouch(null, offsetX, offsetY);
                 this.pointA = null;
                 this.pointB = null;
-            } else if (p.type === PointerEventTypes.POINTERDOWN && srcElement) {
+            } else if (p.type === PointerEventTypes.POINTERDOWN && (srcElement || this._usingNative)) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/BaseCameraPointersInput.ts]: Removed checks for native, only checking for srcElement

File: src/Cameras/Inputs/BaseCameraPointersInput.ts [hunk 4]
```diff
@@ -133,9 +135,9 @@ export abstract class BaseCameraPointersInput implements ICameraInput<Camera> {
                 }
             } else if (p.type === PointerEventTypes.POINTERDOUBLETAP) {
                 this.onDoubleTap(evt.pointerType);
-            } else if (srcElement && (p.type === PointerEventTypes.POINTERUP || (p.type === PointerEventTypes.POINTERMOVE && p.event.button === this._currentActiveButton && this._currentActiveButton !== -1 && !ignoreContext))) {
+            } else if ((srcElement || this._usingNative) && (p.type === PointerEventTypes.POINTERUP || (p.type === PointerEventTypes.POINTERMOVE && p.event.button === this._currentActiveButton && this._currentActiveButton !== -1 && !ignoreContext))) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/BaseCameraPointersInput.ts]: Removed checks for native, only checking for srcElement

File: src/Cameras/Inputs/flyCameraMouseInput.ts [hunk 5]
```diff
@@ -161,9 +163,9 @@ export class FlyCameraMouseInput implements ICameraInput<FlyCamera> {
         var srcElement = <HTMLElement>(e.srcElement || e.target);
 
         // Mouse down.
-        if (p.type === PointerEventTypes.POINTERDOWN && srcElement) {
+        if (p.type === PointerEventTypes.POINTERDOWN && (srcElement || this._usingNative)) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/flyCameraMouseInput.ts]: Removed checks for native, only checking for srcElement

File: src/Cameras/Inputs/flyCameraMouseInput.ts [hunk 6]
```diff
@@ -186,9 +188,9 @@ export class FlyCameraMouseInput implements ICameraInput<FlyCamera> {
             }
         } else
             // Mouse up.
-            if (p.type === PointerEventTypes.POINTERUP && srcElement) {
+            if (p.type === PointerEventTypes.POINTERUP && (srcElement || this._usingNative)) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/flyCameraMouseInput.ts]: Removed checks for native, only checking for srcElement

File: src/Cameras/Inputs/freeCameraMouseInput.ts [hunk 7]
```diff
@@ -85,9 +87,9 @@ export class FreeCameraMouseInput implements ICameraInput<FreeCamera> {
 
                 let srcElement = <HTMLElement>(evt.srcElement || evt.target);
 
-                if (p.type === PointerEventTypes.POINTERDOWN && srcElement) {
+                if (p.type === PointerEventTypes.POINTERDOWN && (srcElement || this._usingNative)) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/freeCameraMouseInput.ts]: Removed checks for native, only checking for srcElement

File: src/Cameras/Inputs/freeCameraMouseInput.ts [hunk 8]
```diff
@@ -110,9 +112,9 @@ export class FreeCameraMouseInput implements ICameraInput<FreeCamera> {
                     if (engine.isPointerLock && this._onMouseMove) {
                         this._onMouseMove(p.event);
                     }
-                } else if (srcElement && (p.type === PointerEventTypes.POINTERUP || (p.type === PointerEventTypes.POINTERMOVE && p.event.button === this._currentActiveButton && this._currentActiveButton !== -1 && !this._usingSafari))) {
+                } else if ((srcElement || this._usingNative) && (p.type === PointerEventTypes.POINTERUP || (p.type === PointerEventTypes.POINTERMOVE && p.event.button === this._currentActiveButton && this._currentActiveButton !== -1 && !this._usingSafari))) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Cameras/Inputs/freeCameraMouseInput.ts]: Removed checks for native, only checking for srcElement

File: src/Helpers/sceneHelpers.ts [hunk 9]
```diff
@@ -144,7 +144,7 @@ Scene.prototype.createDefaultCamera = function (createArcRotateCamera = false, r
         this.activeCamera = camera;
 
         const canvas = this.getEngine().getInputElement();
-        if (attachCameraControls && canvas) {
+        if (attachCameraControls && (canvas || !this.getEngine().getHostDocument())) {
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Helpers/sceneHelpers.ts]: Removed checks for native and canvas

File: src/Inputs/scene.inputManager.ts [hunk 10]
```diff
@@ -482,20 +483,23 @@ export class InputManager {
      */
     public attachControl(attachUp = true, attachDown = true, attachMove = true, elementToAttachTo: Nullable<HTMLElement> = null): void {
         let scene = this._scene;
+        let engine = scene.getEngine();
+
+        if (!engine.getHostDocument()) this._usingNative = true;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Inputs/scene.inputManager.ts]: Removed checks for native

**Review by sebavan — CHANGES_REQUESTED** [review]
Only a couple nits and mainly you need to resolve conflicts  :-) 

I ll merge it in right after.

File: src/Inputs/scene.inputManager.ts [hunk 11]
```diff
@@ -482,20 +482,17 @@ export class InputManager {
      */
     public attachControl(attachUp = true, attachDown = true, attachMove = true, elementToAttachTo: Nullable<HTMLElement> = null): void {
         let scene = this._scene;
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Inputs/scene.inputManager.ts]: fixed nit

File: src/Inputs/scene.inputManager.ts [hunk 12]
```diff
@@ -482,20 +482,17 @@ export class InputManager {
      */
     public attachControl(attachUp = true, attachDown = true, attachMove = true, elementToAttachTo: Nullable<HTMLElement> = null): void {
         let scene = this._scene;
+        let engine = scene.getEngine();
```
[the remark this answers is not in the document]
Comment (PolygonalSun) [reply src/Inputs/scene.inputManager.ts]: fixed nit

**Review by sebavan — APPROVED** [review]

**Pull request merged**